test/form/samples/treeshake-deterministic-dynamic-import/main.js

```diff
@@ -0,0 +1,42 @@
+export async function entry() {
+  // simple
+  const { foo1: foo } = await import('./sub1.js');
+
+  // multiple
+  ;(await import('./sub2.js')).bar2()
+  const { foo2 } = await import('./sub2.js');
+  const { foo3 } = await import('./sub2.js');
+  import('./sub2.js').then(({ baz2 }) => baz2)
+  import('./sub2.js').then(function({ reexported }) { reexported })
+
+  // side-effect only
+  import('./effect-1.js')
+  await import('./effect-2.js')
+  import('./effect-3.js').then(function() { })
+  import('./effect-4.js').then()
+  import('./effect-5.js').catch(() => {})
+  import('./effect-6.js').finally(() => {})
+
+  // bail out
+  const { named1 } = await import('./bail-1.js');
```

Comment:
One last thing that occurred to me would be to double the "bail out" cases as function tests. `form` tests are basically snapshot tests, which means they need to be reviewed thoroughly by a human when anything changes and problems are easily overlooked. Unfortunately, they are the only way to test tree-shaking. `function` tests are better in so far as they actually execute the code. They also get `assert` as a global variable so you can make inline assertions. It could look like this:

```js
// main.js
export const allExports = Promise.all([
  (async () => {
    const value = (await import('./bail-3.js'))[foo];
    assert.strictEqual(value, '...');
  })(),
  ...
]);

// _config.js
...
async exports({allExports}) {
  // to ensure async errors are caught
  await allExports;
},
context: {
  foo: '...'
}
```

Reply:
I found that it would be hard to assert what had been shaken or not in the runtime - so I go with another approach to assert for the generated code with certain marks. Do you think that's a good approach?

Reply:
No, I am not a fan of explicit code assertions. After all, the generated code may change drastically with new Rollup versions while observable effects must not change. Then stick with form tests. But we are bailing out for a reasons. We are not keeping the exports because they are destructured, but because they might actually be used. And asserting their values are right after destructuring is a very simple and effective way of using them. This will also make the test much less artificial.
In `function` tests, you can inject globals via `context`, can await exported promises via `Exports` and have `assert` as a global. This is usually enough material to write tests that are close to what actual users might do and which will never need to be changed if Rollup becomes more advanced.

Reply:
Of course you cannot test what is removed with such a test. For that, you NEED the snapshot/chunking-form test. But the more critical error is if something is removed that we need to keep because that leads to broken code. And for that we have `function` tests.

Reply:
I.e. keep the current test as it is, but add a second function test that just asserts that nothing is removed that is actually used.